Adding CoreDNS integration to Datadog

shraykay opened the pull request:
### What does this PR do?

Adding an integration for CoreDNS -- CoreDNS emits metrics via Prometheus.
See sample events from plugins: [here](http://coredns.io/plugins/metrics), [here](https://coredns.io/plugins/cache/), and [here](https://coredns.io/plugins/proxy/).

### Motivation

Kubernetes is moving from KubeDNS -> CoreDNS starting v1.11

### Review checklist

- [x] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [x] Feature or bugfix has tests
- [x] Git history is clean
- [x] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

First time making a contribution! Please let me know what I may have missed, I borrowed heavily from KubeDNS's integration.

shraykay commented:
@gmmeyer should I make changes to support `OpenMetricsBaseCheck` in this PR or in a separate one?

l0k0ms commented on coredns/metadata.csv:
```diff
@@ -0,0 +1,21 @@
+coredns.response_code_count.count,count,,response code,,instant number of responses per zone and rcode,1,coredns,response code
+coredns.proxy_request_count,gauge,,request,,query count per upstream.,1,coredns,proxy request count
+coredns.proxy_request_count.count,count,,request,,query count per upstream at a moment,1,coredns,proxy request count
+coredns.cache_hits_count,gauge,,hits,,Counter of cache hits by cache type,1,coredns,cache hit
```
Unit name should be `hit` not `hits`

https://docs.datadoghq.com/developers/metrics/#units

l0k0ms commented on coredns/metadata.csv:
```diff
@@ -0,0 +1,21 @@
+coredns.proxy_request_count,gauge,,request,,query count per upstream.,1,coredns,proxy request count
+coredns.proxy_request_count.count,count,,request,,query count per upstream at a moment,1,coredns,proxy request count
+coredns.cache_hits_count,gauge,,hits,,Counter of cache hits by cache type,1,coredns,cache hit
+coredns.cache_hits_count.count,count,,hits,,Counter of cache hits by cache type at a moment,1,coredns,cache hit
```
unit_name should be `hit` not `hits`

https://docs.datadoghq.com/developers/metrics/#units

l0k0ms commented on coredns/metadata.csv:
```diff
@@ -0,0 +1,21 @@
+coredns.proxy_request_count.count,count,,request,,query count per upstream at a moment,1,coredns,proxy request count
+coredns.cache_hits_count,gauge,,hits,,Counter of cache hits by cache type,1,coredns,cache hit
+coredns.cache_hits_count.count,count,,hits,,Counter of cache hits by cache type at a moment,1,coredns,cache hit
+coredns.cache_misses_count,gauge,,misses,,Counter of cache misses.,1,coredns,cache miss
```
r/`misses`/`miss`

l0k0ms commented on coredns/metadata.csv:
```diff
@@ -0,0 +1,21 @@
+coredns.cache_hits_count,gauge,,hits,,Counter of cache hits by cache type,1,coredns,cache hit
+coredns.cache_hits_count.count,count,,hits,,Counter of cache hits by cache type at a moment,1,coredns,cache hit
+coredns.cache_misses_count,gauge,,misses,,Counter of cache misses.,1,coredns,cache miss
+coredns.cache_misses_count.count,count,,misses,,Counter of cache misses at a moment.,1,coredns,cache miss
```
r/`misses`/`miss`

l0k0ms commented on coredns/metadata.csv:
```diff
@@ -0,0 +1,21 @@
+coredns.request_count.count,count,,request,,total query count at a moment,1,coredns,request count
+coredns.request_type_count,gauge,,request type,,counter of queries per zone and type,1,coredns,request type
+coredns.request_type_count.count,count,,request type,,counter of queries per zone and type,1,coredns,request type
+coredns.request_duration.seconds.sum,gauge,,seconds,,duration to process each query,-1,coredns,proxy request duration
```
r/`seconds`/`second`

and for the next 3 metrics too

l0k0ms commented on coredns/metadata.csv:
```diff
@@ -0,0 +1,21 @@
+coredns.proxy_request_duration.seconds.count,count,,seconds,,duration per upstream interaction,-1,coredns,proxy request duration
+coredns.request_size.bytes.sum,gauge,,byte,,size of the request in bytes,0,coredns,request size
+coredns.request_size.bytes.count,count,,byte,,size of the request in bytes,0,coredns,request size
+coredns.cache_size.count,count,,size,,,1,coredns,cache size
```
`size` is not a valid unit_name, you should pick one from this list: https://docs.datadoghq.com/developers/metrics/#units

l0k0ms commented on coredns/README.md:
```diff
@@ -0,0 +1,68 @@
+## Overview
+Get metrics from CoreDNS in real time to:
+
+* Visualize and monitor DNS failures and cache hits/misses
```
No need to have a single bullet point, we should just have one sentence.

l0k0ms commented on coredns/README.md:
```diff
@@ -0,0 +1,68 @@
+Get metrics from CoreDNS in real time to:
+
+* Visualize and monitor DNS failures and cache hits/misses
+## Setup
```
Since this integration is in integration-core. 
The CoreDNS check is packaged with the Agent? if so, no need to manually download `check.py` and `check.conf` files.

l0k0ms commented on coredns/README.md:
```diff
@@ -0,0 +1,68 @@
+5. [Restart the Agent][3] to begin sending CoreDNS metrics to Datadog.
+
+### Validation
+
```
Validation section should follow the same layout as all other integrations in integrations core:
https://github.com/DataDog/integrations-core/tree/master/apache#validation

We don't display the result since it differs between agent v5 and v6.

l0k0ms commented on coredns/README.md:
````diff
@@ -0,0 +1,68 @@
+    [...]
+```
+
+## Compatibility
````
No need to add the compatibility section in the readme. The information is already in the manifest.json file and has a specific display on the public documentation:

https://docs.datadoghq.com/integrations/apache/

l0k0ms commented on coredns/README.md:
```diff
@@ -0,0 +1,68 @@
+See [metadata.csv][5] for a list of metrics provided by this integration.
+
+### Events
+The CoreDNS check does not include any event at this time.
```
We don't need to say "at this time" here. The documentation represents _now_, not a hypothetical future.
> The CoreDNS check does not include any events.

l0k0ms commented on coredns/README.md:
```diff
@@ -0,0 +1,68 @@
+The CoreDNS check does not include any event at this time.
+
+### Service Checks
+The CoreDNS check does not include any service checks at this time.
```
Same comment as for events subsections.

gmmeyer commented:
@shraykay can you do it in this one? That PR should be merged soon!

gmmeyer commented on coredns/tox.ini:
```diff
@@ -0,0 +1,24 @@
+[testenv]
+platform = linux2|darwin
+deps =
+    datadog-checks-base
```
can you change this to `-e../datadog_checks_base[deps]`? That will, I believe, make the test work.

gmmeyer commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,108 @@
+        super(CoreDNSCheck, self).__init__(name, init_config, agentConfig, instances)
+        self.NAMESPACE = 'coredns'
+
+        self.metrics_mapper = {
```
can we make this configurable in the yaml? making this the default and then merging a yaml metrics_mapper map with this one would be a good idea I think

gmmeyer commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,108 @@
+            # https://coredns.io/plugins/cache/
+            'coredns_cache_size': 'cache_size.count',
+        }
+        self.ignore_metrics = {
```
do we wanna ignore all of these? (people can find lots of metrics useful for a lot of reasons)

gzussa commented on coredns/requirements-dev.txt:
```diff
@@ -0,0 +1,2 @@
+mock==2.0.0
```
You can just import `datadog-checks-dev` instead of those two

gzussa commented on coredns/setup.py:
```diff
@@ -0,0 +1,61 @@
+    packages=['datadog_checks.coredns'],
+
+    # Run-time dependencies
+    install_requires=get_requirements('requirements.in')+[
```
You can remove the `get_requirements('requirements.in')` since it has nothing in it.

gzussa commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@pytest.fixture
+def aggregator():
```
You can remove this function since it is part of `datadog_check_base`. 
You can look at the `Vault` integrations for an example.

BrianChristie commented:
Thanks for this work!

@ datadog folks, it would be great to merge this soon because as you can see in the Kubernetes V1.11 release notes https://kubernetes.io/docs/setup/release/notes/#sig-cluster-lifecycle , CoreDNS has replaced kube-dns as the default DNS provider so all new clusters have only CoreDNS running.

shraykay commented:
i will work on this today/tomorrow! will have an updated version for you soon.

shraykay commented:
I pushed a newer version, however I'm having trouble running an integration test using docker-compose. can anyone provide some pointers?

gzussa commented on coredns/tox.ini:
```diff
@@ -0,0 +1,37 @@
+
+[common]
+deps =
+    ../datadog_checks_base
```
You should be able to replace both of those lines with `-e../datadog_checks_base[deps]`

gzussa commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,170 @@
+    }
+
+@pytest.fixture
+def aggregator():
```
aggregator is globally available with `datadog_checks_dev` [there](https://github.com/DataDog/integrations-core/blob/master/datadog_checks_dev/datadog_checks/dev/plugin/plugin.py#L10) so you don't need to add it here :)

gzussa commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,170 @@
+
+@pytest.fixture
+def coredns():
+    env = os.environ
```
instead of calling subprocess command to start your container, you can leverage utilities in this [file](https://github.com/DataDog/integrations-core/blob/master/datadog_checks_dev/datadog_checks/dev/docker.py#L46). You can find good examples [here](https://github.com/DataDog/integrations-core/blob/master/zk/tests/conftest.py#L111) or [there](https://github.com/DataDog/integrations-core/blob/master/sqlserver/tests/conftest.py#L81). Feel free to take a look at what the `datadog_checks_dev` package contains since it holds a lot of useful methods :)

shraykay commented:
@gzussa thanks! was able to get docker-compose going with your help, could you help me figure out why flake8 isn't running? the error seems a bit vague

ofek commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,110 @@
+
+        return instance
+
+    def submit_as_gauge_and_monotonic_count(self, metric_suffix, message, scraper_config):
```
I think these should only be submitted as `monotonic_count`. See https://github.com/DataDog/integrations-core/issues/1303

ofek replied:
cc @mfpierre to verify

mfpierre replied:
I agree we should use the `send_monotonic_counter` to `True` here I double-checked all the affected metrics are of the `counter` type.
As a result we can remove all the custom functions and `METRIC_TRANSFORMERS`

ofek commented on coredns/setup.py:
```diff
@@ -0,0 +1,51 @@
+# Get the long description from the README file
+with open(path.join(HERE, 'README.md'), encoding='utf-8') as f:
+    long_description = f.read()
+
```
Need to define `CHECKS_BASE_REQ = 'datadog-checks-base'`

ofek commented on coredns/setup.py:
```diff
@@ -0,0 +1,51 @@
+    keywords='datadog agent check',
+    url='https://github.com/DataDog/integrations-core',
+    author='CoreDNS',
+    author_email='shray@namely.com',
```
```
author='Datadog',
author_email='packages@datadoghq.com',
```

ofek commented on coredns/setup.py:
```diff
@@ -0,0 +1,51 @@
+
+    # Extra files to ship with the wheel package
+    package_data={'datadog_checks.coredns': ['conf.yaml.example']},
+    include_package_data=True,
```
```
    # The package we're going to ship
    packages=['datadog_checks', 'datadog_checks.vault'],

    # Run-time dependencies
    install_requires=[CHECKS_BASE_REQ],

    # Extra files to ship with the wheel package
    include_package_data=True,
```

ofek commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,172 @@
+
+
+@pytest.fixture(scope="session")
+def spin_up_coredns():
```
Can you please put all these fixtures in a `conftest.py`? At that point they will be globally available.

ofek commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,172 @@
+                res = requests.get(URL)
+                # create some metrics by using dig
+                subprocess.check_call(DIG_ARGS, env=env)
+                res.raise_for_status
```
`res.raise_for_status()`

ofek commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,172 @@
+
+    compose_file = os.path.join(HERE, 'docker', 'docker-compose.yml')
+    env = os.environ
+    env['COREDNS_CONFIG_FOLDER'] = CONFIG_FOLDER
```
`env = {'COREDNS_CONFIG_FOLDER': CONFIG_FOLDER}`

ofek commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,172 @@
+    with open(mesh_file_path, 'rb') as f:
+        text_data = f.read()
+
+    p = mock.patch('requests.get', return_value=MockResponse(text_data, 'text/plain; version=0.0.4'), __name__='get')
```
```
with mock.patch(...):
    yield
```

rishabh commented on coredns/datadog_checks/coredns/data/conf.yaml.example:
```diff
@@ -0,0 +1,10 @@
+
+instances:
+    # url of the metrics endpoint of prometheus
+  - prometheus_endpoint: http://localhost:9153/metrics
```
Can you please use `prometheus_url` instead of `prometheus_endpoint`?

rishabh commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,110 @@
+        """
+        Set up coredns instance so it can be used in OpenMetricsBaseCheck
+        """
+        endpoint = instance.get('prometheus_endpoint')
```
This should also be `prometheus_url`.

rishabh commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,110 @@
+        """
+        endpoint = instance.get('prometheus_endpoint')
+        if endpoint is None:
+            raise CheckException("Unable to find prometheus endpoint in config file.")
```
And we can just change this to `prometheus_url`.

rishabh commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,110 @@
+        super(CoreDNSCheck, self).__init__(name, init_config, agentConfig, instances=generic_instances)
+
+    def check(self, instance):
+        endpoint = instance.get('prometheus_endpoint')
```
And here as well `prometheus_url`.

rishabh commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,110 @@
+            # Primarily, metrics are emitted by the prometheus plugin: https://coredns.io/plugins/metrics/
+            # Note: the count metrics were moved to specific functions
+            # below to be submitted as both gauges and monotonic_counts
+            # 'coredns_dns_response_rcode_count_total': self.coredns_dns_response_rcode_count_total,
```
Since we're not using the transformer functions, we can just get rid of this dictionary.

rishabh commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,110 @@
+    def check(self, instance):
+        endpoint = instance.get('prometheus_endpoint')
+        scraper_config = self.config_map[endpoint]
+        self.process(scraper_config, metric_transformers=self.METRIC_TRANSFORMERS)
```
And then we can also get rid of this named argument.

rishabh commented on coredns/datadog_checks/coredns/data/auto_conf.yaml:
```diff
@@ -0,0 +1,11 @@
+instances:
+  # To enable CoreDNS metrics you must specify the prometheus url and enable the plugin within coredns
+  # See: https://coredns.io/plugins/metrics/
+  - prometheus_endpoint: "http://%%host%%:9153/metrics"
```
And this should also be `prometheus_url`.

rishabh commented on coredns/README.md:
```diff
@@ -0,0 +1,66 @@
+  annotations:
+    ad.datadoghq.com/coredns.check_names: '["coredns"]'
+    ad.datadoghq.com/coredns.init_configs: '[{}]'
+    ad.datadoghq.com/coredns.instances: '[[{"prometheus_endpoint":"http://%%host%%:9153/metrics", "tags":["dns-pod:%%host%%"]}]]'
```
And this one, too :)

rishabh commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,172 @@
+from datadog_checks.utils.common import get_docker_hostname
+
+instance = {
+    'prometheus_endpoint': 'http://localhost:9153/metrics',
```
Let's fix the tests as well to use `prometheus_url`.

rishabh commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,172 @@
+@pytest.fixture
+def dockerinstance():
+    return {
+        'prometheus_endpoint': URL,
```
And let's also fix the docker instance config to use `prometheus_url`.

mfpierre commented on coredns/tests/test_check.py:
```diff
@@ -0,0 +1,172 @@
+        NAMESPACE + '.proxy_request_duration.seconds.sum',
+        NAMESPACE + '.cache_size.count',
+    ]
+    COUNT_METRICS = [
```
we should remove this empty list

shraykay commented:
@gmmeyer I parameterized the metrics and added the go_metrics by default.

@rishabh @ofek  I think I've addressed your concerns, can you take a look?

thanks!

devonboyer commented on coredns/datadog_checks/coredns/data/auto_conf.yaml:
```diff
@@ -0,0 +1,35 @@
+ad_identifiers:
+  - coredns/coredns
```
The `ad_identifier` identifier for auto_conf needs to be the short name for container images. I think this needs to be

```
ad_identifiers:
  - coredns
```

ofek commented on coredns/manifest.json:
```diff
@@ -0,0 +1,21 @@
+  "supported_os": [
+    "linux"
+  ],
+  "guid": "000-000",
```
@shraykay

```
$ python -c "from uuid import uuid4; print(uuid4())"
9b316155-fc8e-4cb0-8bd5-8af270759cfb
```

ofek commented on coredns/setup.py:
```diff
@@ -0,0 +1,54 @@
+    packages=['datadog_checks', 'datadog_checks.coredns'],
+
+    # Extra files to ship with the wheel package
+    package_data={'datadog_checks.coredns': ['conf.yaml.example']},
```
You can remove this

ofek commented on coredns/setup.py:
```diff
@@ -0,0 +1,54 @@
+
+    # Run-time dependencies
+    install_requires=[CHECKS_BASE_REQ, ],
+    setup_requires=['pytest-runner', ],
```
You can remove `setup_requires`

shraykay commented on coredns/datadog_checks/coredns/coredns.py:
```diff
@@ -0,0 +1,127 @@
+
+        metrics = instance.get('metrics', [])
+        if metrics is None:
+            metrics = metrics + [GO_METRICS]
```
ah okay yeah, I am making the assumption that CoreDNS users will use at minimum the default plugins out of the box that emit the stats mentioned there.  Would a field called `additional_metrics` for any extra CoreDNS metrics suffice? @rishabh

rishabh replied:
I apologize if I'm misunderstanding, but I think you don't need to specify an extra field; you can continue to use `metrics`.

What my idea is that we always send `DEFAULT_METRICS` and `GO_METRICS` no matter what the user defines, and then whatever they specify in `metrics` are the additional metrics we collect.

I'm open to ideas :)

gzussa commented:
Sorry i didn't see you message earlier, I am glad my comments could help. 💯 
This looks great. 

Other than currently existing feedback i don't have much to say. 👍 

- Regarding config file make sure to follow this [spec](https://github.com/DataDog/integrations-core/blob/master/docs/dev/new_check_howto.md#configuration-file).
- Regarding the manifest and the metadata file make sure to [read this doc](https://github.com/DataDog/integrations-core/blob/master/docs/dev/new_check_howto.md#metadata) otherwise the CI will fail when we will merge to master.

shraykay commented:
greetings! just following up on this, is there anything else required or can we get this merged in? :)
